java/client/src/main/java/com/vesoft/nebula/client/graph/storage/general/GeneralStorageClient.java

```
@@ -0,0 +1,659 @@
+            threadPool.submit(() -> {
+                if (doPut(spaceName, entry.getKey(), entry.getValue())) {
+                    responses.add(true);
+                } else {
+                    responses.add(false);
+                }
+                countDownLatch.countDown();
+            });
+        }
+
+        try {
+            countDownLatch.await();
+        } catch (InterruptedException interruptedE) {
+            LOGGER.error("Put interrupted:", interruptedE);
+            return false;
+        }
+
+        for (Boolean resp : responses) {
+            if (!resp) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+
+    /**
+     * wrap request and execute put
+     *
+     * @param spaceName nebula graph space
+     * @param leader    leader for part
+     * @param request   PutRequest
+     * @return boolean
+     */
+    private boolean doPut(String spaceName, HostAndPort leader, KVPutRequest request) {
+        // there is at least one retry for leader change.
+        int retry = 0;
+        while (retry++ != metaInfo.getConnectionRetry()) {
+
+            try {
+                connection = getConnection(leader);
+                if (connection == null) {
+                    LOGGER.error("connection is in usage");
+                    return false;
+                }
+                connection.setBusy();
+            } catch (Exception e) {
+                LOGGER.error("Connect error: ", e);
+                return false;
+            }
+
+            try {
+                ExecResponse result = connection.put(request);
+                if (isSuccessfully(result)) {
+                    connection.release();
+                    return true;
+                }
+                for (PartitionResult partResult : result.getResult().getFailed_parts()) {
+                    if (partResult.code == ErrorCode.E_LEADER_CHANGED) {
```

if the error code is not leader_change, directly return, no need to retry. and better to return the error msg to users